HidPkg/Crates/HiiKeyboardLayout/src/lib.rs

```diff
@@ -0,0 +1,782 @@
+  /// driver. See [`r_efi::protocols::hii_database`] for modifier definitions.
+  pub modifier: u16,
+  /// Indicates what modifiers affect this key. See [`r_efi::protocols::hii_database`] for "affected by" definitions.
+  pub affected_attribute: u16,
+}
+
+/// Non-Spacing HII Key Descriptor variant. Used for "non-spacing" keys.
+/// Refer to discussion in UEFI spec version 2.10 section 33.2.4.3 for information on "non-spacing" keys and how they
+/// are used.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct HiiNsKeyDescriptor {
+  /// The descriptor for the "non-spacing key" itself.
+  pub descriptor: HiiKeyDescriptor,
+  /// The list of descriptors that are active if the "non-spacing" key has been pressed.
+  pub dependent_keys: Vec<HiiKeyDescriptor>,
+}
+
+/// HII Key descriptor enumeration.
+/// HII spec allows for two types of key descriptors - normal and "non-spacing".
+/// Refer to UEFI spec version 2.10 section 33.2.4.3
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum HiiKey {
+  Key(HiiKeyDescriptor),
+  NsKey(HiiNsKeyDescriptor),
+}
+
+/// Enumeration of physical keys.
+/// Refer to UEFI spec version 2.10 section 34.8.10 and section 33.2.4.1.
+#[rustfmt::skip]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+//Note: UEFI specifies this as an C enum. That means the size is a bit ambiguous; but most compilers
+//will make it 32-bit, so that's what this implementation assumes.
+#[repr(u32)]
+pub enum EfiKey {
+  EfiKeyLCtrl, EfiKeyA0, EfiKeyLAlt, EfiKeySpaceBar, EfiKeyA2, EfiKeyA3, EfiKeyA4, EfiKeyRCtrl, EfiKeyLeftArrow,
+  EfiKeyDownArrow, EfiKeyRightArrow, EfiKeyZero, EfiKeyPeriod, EfiKeyEnter, EfiKeyLShift, EfiKeyB0, EfiKeyB1, EfiKeyB2,
+  EfiKeyB3, EfiKeyB4, EfiKeyB5, EfiKeyB6, EfiKeyB7, EfiKeyB8, EfiKeyB9, EfiKeyB10, EfiKeyRShift, EfiKeyUpArrow,
+  EfiKeyOne, EfiKeyTwo, EfiKeyThree, EfiKeyCapsLock, EfiKeyC1, EfiKeyC2, EfiKeyC3, EfiKeyC4, EfiKeyC5, EfiKeyC6,
+  EfiKeyC7, EfiKeyC8, EfiKeyC9, EfiKeyC10, EfiKeyC11, EfiKeyC12, EfiKeyFour, EfiKeyFive, EfiKeySix, EfiKeyPlus,
+  EfiKeyTab, EfiKeyD1, EfiKeyD2, EfiKeyD3, EfiKeyD4, EfiKeyD5, EfiKeyD6, EfiKeyD7, EfiKeyD8, EfiKeyD9, EfiKeyD10,
+  EfiKeyD11, EfiKeyD12, EfiKeyD13, EfiKeyDel, EfiKeyEnd, EfiKeyPgDn, EfiKeySeven, EfiKeyEight, EfiKeyNine, EfiKeyE0,
+  EfiKeyE1, EfiKeyE2, EfiKeyE3, EfiKeyE4, EfiKeyE5, EfiKeyE6, EfiKeyE7, EfiKeyE8, EfiKeyE9, EfiKeyE10, EfiKeyE11,
+  EfiKeyE12, EfiKeyBackSpace, EfiKeyIns, EfiKeyHome, EfiKeyPgUp, EfiKeyNLck, EfiKeySlash, EfiKeyAsterisk, EfiKeyMinus,
+  EfiKeyEsc, EfiKeyF1, EfiKeyF2, EfiKeyF3, EfiKeyF4, EfiKeyF5, EfiKeyF6, EfiKeyF7, EfiKeyF8, EfiKeyF9, EfiKeyF10,
+  EfiKeyF11, EfiKeyF12, EfiKeyPrint, EfiKeySLck, EfiKeyPause, EfiKeyIntl0, EfiKeyIntl1, EfiKeyIntl2, EfiKeyIntl3,
+  EfiKeyIntl4, EfiKeyIntl5, EfiKeyIntl6, EfiKeyIntl7, EfiKeyIntl8, EfiKeyIntl9,
+}
+
+impl TryFrom<u32> for EfiKey {
+  type Error = &'static str;
+  fn try_from(value: u32) -> Result<Self, Self::Error> {

```

Comment:
This is a common enough pattern (converting between an enum variant and integer), should we go ahead and bring in the [`num_derive`](https://docs.rs/num-derive/0.4.1/num_derive/#derives) crate and then `#[derive(FromPrimitive)]` on the enum?

Reply:
That should result in an `Option` and be able to avoid the unsafe transmute here.

Reply:
happy to do that, it would be cleaner. But my initial intuition was to avoid the dependency.

Reply:
I thought the same when I ran into this, but I think we're inevitably going to keep having to do this given the code we're writing so it's a useful dependency.

Reply:
Implemented suggestion.